pyscf/sgx/sgx.py

```diff
@@ -109,26 +119,48 @@ def reset(self, mol=None):
             return mf_class.reset(self, mol)
 
         def pre_kernel(self, envs):
-            self._in_scf = True
+            self.direct_scf = False # should always be False
+            if self.with_df.grids_level_i != self.with_df.grids_level_f:
+                self._in_scf = True
+            if self.with_df.pjs:
+                self.with_df.dfj = True # no SGX-J allowed if P-junction screening on
 
         def get_jk(self, mol=None, dm=None, hermi=1, with_j=True, with_k=True,
                    omega=None):
             if dm is None: dm = self.make_rdm1()
             with_df = self.with_df
             if not with_df:
                 return mf_class.get_jk(self, mol, dm, hermi, with_j, with_k, omega)
+            if self.opt is None and self.with_df.direct_j and (not self.with_df.dfj):
+                self.opt = self.init_direct_scf(mol)
 
             if self._in_scf and not self.direct_scf:
                 if numpy.linalg.norm(dm - self._last_dm) < with_df.grids_switch_thrd:
                     logger.debug(self, 'Switching SGX grids')
                     with_df.build(level=with_df.grids_level_f)
                     self._in_scf = False
                     self._last_dm = 0
-                else:
-                    self._last_dm = numpy.asarray(dm)
+                    self._last_vj = 0
+                    self._last_vk = 0
 
-            return with_df.get_jk(dm, hermi, with_j, with_k,
-                                  self.direct_scf_tol, omega)
+            if self.direct_scf_sgx:
+                vj, vk = with_df.get_jk(dm-self._last_dm, hermi, self.opt,
+                                        with_j, with_k,
+                                        self.direct_scf_tol, omega)
+
+                vj += self._last_vj
+                vk += self._last_vk
+
+                self._last_dm = numpy.asarray(dm)
+                self._last_vj = vj.copy()
+                self._last_vk = vk.copy()
+
+            else:
+                self._last_dm = numpy.asarray(dm)
+                vj, vk = with_df.get_jk(dm, hermi, self.opt, with_j, with_k,
+                                        self.direct_scf_tol, omega)
+
+            return vj, vk
 
         def post_kernel(self, envs):
```

Comment:
Also reset last_vj and last_vk in post_kernel

Reply:
Fixed by 4ef2ea7